src/workerd/io/trace-common.h

```diff
@@ -220,30 +359,192 @@ struct Log final {
   // TODO(soon): Just string for now.  Eventually, capture serialized JS objects.
   kj::String message;
 
-  void copyTo(rpc::Trace::Log::Builder builder);
+  void copyTo(rpc::Trace::Log::Builder builder) const;
+  Log clone() const;
+};
+
+struct LogV2 final {
+  explicit LogV2(
+      kj::Date timestamp, LogLevel logLevel, kj::Array<kj::byte> data, Tags tags = nullptr);
+  LogV2(rpc::Trace::LogV2::Reader reader);
+  LogV2(LogV2&&) = default;
+  LogV2& operator=(LogV2&&) = default;
+  KJ_DISALLOW_COPY(LogV2);
+
+  // Only as accurate as Worker's Date.now(), for Spectre mitigation.
+  kj::Date timestamp;
+
+  LogLevel logLevel;
+  kj::Array<kj::byte> data;
+  Tags tags = nullptr;
+
+  void copyTo(rpc::Trace::LogV2::Builder builder) const;
+  LogV2 clone() const;
 };
 
 struct Exception final {
-  explicit Exception(
-      kj::Date timestamp, kj::String name, kj::String message, kj::Maybe<kj::String> stack);
+  struct Detail {
+    kj::Maybe<kj::Own<Exception>> cause = kj::none;
+    kj::Array<kj::Own<Exception>> errors = nullptr;
+    bool remote = false;
+    bool retryable = false;
+    bool overloaded = false;
+    bool durableObjectReset = false;
+    Tags tags = nullptr;
+
+    Detail clone() const;
+  };
+
+  explicit Exception(kj::Date timestamp,
+      kj::String name,
+      kj::String message,
+      kj::Maybe<kj::String> stack,
+      kj::Maybe<Detail> detail = kj::none);
   Exception(rpc::Trace::Exception::Reader reader);
   Exception(Exception&&) = default;
+  Exception& operator=(Exception&&) = default;
   KJ_DISALLOW_COPY(Exception);
-  ~Exception() noexcept(false) = default;
 
   // Only as accurate as Worker's Date.now(), for Spectre mitigation.
   kj::Date timestamp;
 
   kj::String name;
   kj::String message;
-
   kj::Maybe<kj::String> stack;
+  Detail detail;
+
+  void copyTo(rpc::Trace::Exception::Builder builder) const;
+  Exception clone() const;
+};
+
+struct Subrequest final {
```

Comment:
Would you mind adding a comment to every struct you created? It's *getting* hard to review.